### DISCUSSION TITLE
deafult DEB repository to override default one(s)

### DISCUSSION THREAD

**Discussion by artynet** [termux/termux-app#3609] [General]
Hello folks,

is there a way to edit the default APT binary entry so to point that to a custom debian repository (i.e. nexus3 hosted). I switched it here:

https://github.com/termux/termux-packages/blob/25498b08a581d66d701543aeaf35dad53ffe8c53/scripts/generate-bootstraps.sh#L26

but the i took for granted that the entry will propagate on the final APK build of termux-app itself. Any pointers to make it ? 

I would really appreciate it, thanks in advance.

Arturo

**Comment by TomJo2000** [comment]
There is no modification of the binary required to change the repos.
You already linked the line in the `generate-bootstrap.sh` script that sets the repo URL.

Changing the URL there and building the APK will propagate the default mirror to be that URL.

---

However.
### You don't even need to rebuild the APK, if you want to manually change the mirror.

You can simply edit the:
`/data/data/com.termux/files/usr/etc/apt/sources.list`
file to point at your own repo(s).

Please note Termux's repos are not Debian's repos.
Our packages are compiled to expect being placed in `/data/data/com.termux/files/usr/bin` in order to comply with Android's execution limitations, regular mainline Debian/Ubuntu packages are most likely not going to work on Termux.
Use of the `.deb` packaging format by Termux is for convenience and established tooling only.

**Reply by artynet** [reply]
Hello @TomJo2000,

thank you for you quick answer. Altough I changed the URL in the script, it did not propagate trough the APK building with android studio. As long as it concerns the package structure, of course I have built out-of-the-tree additional packages that I need for my custom build using the **termux-packages** SDK.

Another issue that popped out concerns the gpg key for the APT repository.....do i have to encrypt the public GPG key before adding it to the **termux-keyring** package ?

**Reply by TomJo2000** [reply]
Our package repositories are PGP signed to validate package integrity;
ensuring the package has not been corrupted in transit or been otherwise tampered with.

You can make a signing exception for your repository by putting `[trusted=yes]` in the repository line.
e.g.
```yaml
deb [trusted=yes] https://my.repo.url/termux/apt/termux-main stable main
```

### (Please be again advised that the GPG/PGP signature check of packages is both a security and integrity measure and disabling it for repositories you do not control is highly discouraged.)

---

If you wish to just test some packages
you have built locally you can instead install the `.deb` files directly from local sources using:
 `apt install /path/to/package.deb`

**Reply by artynet** [reply]
this is the error I get when i run the command `pkg update`:

```
Checking availability of current mirror:
[*] https://MY-CUSTOM-REPOSITORY: ok
Get:1 https://MY-CUSTOM-REPOSITORY stable InRelease [2057 B]
Err:1 https://MY-CUSTOM-REPOSITORY stable InRelease                                                                    
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 04BA8B8D9ACEAC0E
Reading package lists... Done                                                                                                                    
W: https://MY-CUSTOM-REPOSITORY/dists/stable/InRelease: The key(s) in the keyring /data/data/com.termux/files/usr/etc/apt/trusted.gpg.d/my-custom-repository.gpg are ignored as the file has an unsupported filetype.
W: GPG error: https://MY-CUSTOM-REPOSITORY stable InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 04BA8B8D9ACEAC0E
E: The repository 'https://MY-CUSTOM-REPOSITORY stable InRelease' is not signed.
N: Metadata integrity can't be verified, repository is disabled now.
N: Possible causes: unstable or tampered Internet connection, wrong sources.list, outdated keyring or host is down currently.
N: Please note that all hosting problems or other serious issues we announce on our social media pages.
N: Use termux-change-repo for switching to a mirror.
```
**The key(s) in the keyring /data/data/com.termux/files/usr/etc/apt/trusted.gpg.d/my-custom-repository.gpg are ignored as the file has an unsupported filetype.**

As I am used to work with, I usually create my nexus3 deb repository with a pair of GPG keys, adding the private one to nexus and the public one to the linux distro (in this case termux). When opening the gpg keys in the **termux-keyring** folder they seem to be in binary format, any pointers on how to get a similar binary for my gpg key ?

**Reply by agnostic-apollo** [reply]
Simple `gpg ---export` should do it in binary format.

https://www.gnupg.org/gph/en/manual/x56.html

Default repo is set by `termux-tools` package at https://github.com/termux/termux-tools/blob/master/mirrors/default, you can either build it yourself and publish to your repo or modify the respective file at end of `generate-bootstraps` before `create_bootstrap_archive`. The `REPO_BASE_URL` is meant for where to download the packages from.

**Reply by artynet** [reply]
Thank you @agnostic-apollo,

that's exactly what I wanted to know. I usually export the GPG pulic key with the `--armor` additional option so to make it readable (as custom use with apt-based repositories). Once the key has been replaced, I was able to perform the `pkg update` refresh from my repository wythout any error.

As far as it concerns the `termux-tools` repository, the best solution for me is to apply a local patch with my default URL. I have really appreciated your help.....kind regards ;)

**Reply by agnostic-apollo** [reply]
You are welcome. Glad it worked out.
